### ISSUE TITLE
Creating DMatrix from pyarrow.Table vs. pandas.DataFrame

### ISSUE THREAD

**Issue by AlenkaF** [dmlc/xgboost#9692]
We have been looking at the interoperability between Arrow data nad XGBoost through the use of PyArrow implementation and noticed that when constructing `xgb.DMatrix` there is a big performance difference between consuming data from PyArrow Table or a pandas dataframe.

I saw a different benchmark added to one of the the implementations description https://github.com/dmlc/xgboost/pull/7283#issue-1014180964 so I am curious how this is so different from what I am seeing.

Example: with a small dataset ([DryBeanDataset](https://archive.ics.uci.edu/dataset/602/dry+bean+dataset)) saved in a parquet file format we read the data with pyarrow and pandas (which uses pyarrow under the hood https://pandas.pydata.org/docs/reference/api/pandas.read_parquet.html) and constructed the `xgb.DMatrix` from there.

Using `memory_profiler` to check memory consumption line-by-line and `timeit` to time only the line where the `xgb.DMatrix` is constructed, we can see time to convert a pandas dataframe is much faster then converting a pyarrow table on my M1:

```
Time to convert pyarrow table: 3.3553183339972747
Filename: xgboost-profile.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     6     43.5 MiB     43.5 MiB           1   @profile
     7                                         def run_pyarrow():
     8     43.5 MiB      0.0 MiB           1       mysetup = '''
     9                                         import pyarrow as pa
    10                                         import pyarrow.parquet as pq
    11                                         import xgboost as xgb
    12                                         table = pq.read_table("df_parquet.parquet")
    13                                         table = table.drop_columns("Class")'''
    14                                         
    15     43.5 MiB      0.0 MiB           1       mycode = "xgb.DMatrix(table)"
    16                                         
    17    128.0 MiB     84.5 MiB           2       time = timeit.timeit(setup=mysetup,
    18     43.5 MiB      0.0 MiB           1                            stmt=mycode,
    19     43.5 MiB      0.0 MiB           1                            number=1)
    20    128.1 MiB      0.1 MiB           1       print('Time to convert pyarrow table: {}'.format(time))

Time to convert pandas df: 0.005939499998930842
Filename: xgboost-profile.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    22     42.6 MiB     42.6 MiB           1   @profile
    23                                         def run_pandas():
    24     42.6 MiB      0.0 MiB           1       mysetup = '''
    25                                         import pandas as pd
    26                                         import xgboost as xgb
    27                                         df = pd.read_parquet("df_parquet.parquet")
    28                                         df = df.drop(columns=['Class'])'''
    29                                         
    30     42.6 MiB      0.0 MiB           1       mycode = "xgb.DMatrix(df)"
    31                                         
    32    168.0 MiB    125.4 MiB           2       time = timeit.timeit(setup=mysetup,
    33     42.6 MiB      0.0 MiB           1                            stmt=mycode,
    34     42.6 MiB      0.0 MiB           1                            number=1)
    35                                             
    36    168.0 MiB      0.0 MiB           1       print('Time to convert pandas df: {}'.format(time))
```

is this due to the layout of the data (2d vs multiple 1D structures)?

Also interesting behavior when doing 1000 executions of `xgb.DMatrix(df)`, the time between pyarrow and pandas
becomes closer pointing to a possible internal caching in pyarrow case?

```
Time to convert pyarrow table: 7.834925041999668
Filename: xgboost-profile.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     6     43.9 MiB     43.9 MiB           1   @profile
     7                                         def run_pyarrow():
     8     43.9 MiB      0.0 MiB           1       mysetup = '''
     9                                         import pyarrow as pa
    10                                         import pyarrow.parquet as pq
    11                                         import xgboost as xgb
    12                                         table = pq.read_table("df_parquet.parquet")
    13                                         table = table.drop_columns("Class")'''
    14                                         
    15     43.9 MiB      0.0 MiB           1       mycode = "xgb.DMatrix(table)"
    16                                         
    17    138.4 MiB     94.5 MiB           2       time = timeit.timeit(setup=mysetup,
    18     43.9 MiB      0.0 MiB           1                            stmt=mycode,
    19     43.9 MiB      0.0 MiB           1                            number=1000)
    20    138.5 MiB      0.1 MiB           1       print('Time to convert pyarrow table: {}'.format(time))

Time to convert pandas df: 4.570458542002598
Filename: xgboost-profile.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    22     43.8 MiB     43.8 MiB           1   @profile
    23                                         def run_pandas():
    24     43.8 MiB      0.0 MiB           1       mysetup = '''
    25                                         import pandas as pd
    26                                         import xgboost as xgb
    27                                         df = pd.read_parquet("df_parquet.parquet")
    28                                         df = df.drop(columns=['Class'])'''
    29                                         
    30     43.8 MiB      0.0 MiB           1       mycode = "xgb.DMatrix(df)"
    31                                         
    32    173.3 MiB    129.5 MiB           2       time = timeit.timeit(setup=mysetup,
    33     43.8 MiB      0.0 MiB           1                            stmt=mycode,
    34     43.8 MiB      0.0 MiB           1                            number=1000)
    35                                             
    36    173.3 MiB      0.0 MiB           1       print('Time to convert pandas df: {}'.format(time))
```

The code I used for this example:

```
import sys
import timeit
from memory_profiler import profile


@profile
def run_pyarrow():
    mysetup = '''
import pyarrow as pa
import pyarrow.parquet as pq
import xgboost as xgb
table = pq.read_table("df_parquet.parquet")
table = table.drop_columns("Class")'''

    mycode = "xgb.DMatrix(table)"

    time = timeit.timeit(setup=mysetup,
                         stmt=mycode,
                         number=1)
    print('Time to convert pyarrow table: {}'.format(time))

@profile
def run_pandas():
    mysetup = '''
import pandas as pd
import xgboost as xgb
df = pd.read_parquet("df_parquet.parquet")
df = df.drop(columns=['Class'])'''

    mycode = "xgb.DMatrix(df)"

    time = timeit.timeit(setup=mysetup,
                         stmt=mycode,
                         number=1)
    
    print('Time to convert pandas df: {}'.format(time))


if __name__ == '__main__':
    case = sys.argv[1]
    if case == "pyarrow":
        run_pyarrow()
    elif case == "pandas":
        run_pandas()

```

and the versions of the packages:

```
pandas            2.0.3
pyarrow           13.0.0
xgboost           2.0.0
```

I know benchmarks like this can be tricky to look at and not always correct but I would be grateful if a discussion about possible improvements for Arrow data support could be opened.

**Comment by trivialfis** [comment]
Hi, the original integration of arrow was removed recently (after 2.0, coming in 2.1), and now it's the same as the [arrow-backed pandas dataframe](https://github.com/dmlc/xgboost/blob/ea9f09716bd426f235ce785d6286e0ad26350ce1/python-package/xgboost/testing/data.py#L190).

**Comment by trivialfis** [comment]
The PR was here: https://github.com/dmlc/xgboost/pull/9612 . Going through the pandas route makes most of the feature implementations much easier. In the future, I hope that the dataframe protocol can be used instead.

**Comment by AlenkaF** [comment]
Thank you for the quick response @trivialfis !

Too bad you are not using C Data Interface anymore. I understand the decreased mantainence burden in going through pandas (numpy) but there is a lot of potential in using C Data Interface. Specially for all the projects using Arrow data (DuckDB, Polars, etc.).

Dataframe interchange protocol is also a possibility, yes. Again with similar thought as above.
We are also working in pyarrow to produce PyCapsule as defined by the DLPack and that will be another option also.

All above would be better solutions or alternatives (C Data Interface included) as there are [not needed copies being made in the new path](https://github.com/dmlc/xgboost/blob/3b86260b506154762f3fbd3daaf71c08843ecaf0/python-package/xgboost/data.py#L476-L478) when going through pandas. I would try for zero-copy in case of Arrow data.

As for the example I have added in the issue, after reviewing it with @jorisvandenbossche we realised the huge diff in the first example must be due to the first timeit call needing some extra imports/setup.

Making the issue simpler as the memory profiler is not really needed (though I like the extra info) and using magics in iPython, then the difference between pyarrow and pandas is not that big. On my machine (the diff is even smaller when run on Joris’ machine):

```python
In [9]: import pyarrow as pa
   ...: import pyarrow.parquet as pq
   ...: import xgboost as xgb
   ...: 
   ...: table = pq.read_table(“df_parquet.parquet”)
   ...: table = table.drop_columns(“Class”)
   ...: 
   ...: import pandas as pd
   ...: df = pd.read_parquet(“df_parquet.parquet”)
   ...: df = df.drop(columns=[‘Class’])

In [10]: xgb.__version__
Out[10]: ‘2.0.0’

In [11]: %timeit xgb.DMatrix(df)
1.81 ms ± 27.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [12]: %timeit xgb.DMatrix(table)
2.31 ms ± 18.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

and the time is a bit bigger in both cases if using the dev version of XGBoost:

```python
In [1]: import pyarrow as pa
   ...: import pyarrow.parquet as pq
   ...: import xgboost as xgb
   ...: 
   ...: table = pq.read_table(“df_parquet.parquet”)
   ...: table = table.drop_columns(“Class”)
   ...: 
   ...: import pandas as pd
   ...: df = pd.read_parquet(“df_parquet.parquet”)
   ...: df = df.drop(columns=[‘Class’])

In [2]: xgb.__version__
Out[2]: ‘2.1.0-dev’

In [3]: %timeit xgb.DMatrix(df)
2.28 ms ± 88.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [4]: %timeit xgb.DMatrix(table)
3.83 ms ± 54.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by trivialfis** [comment]
Thank you for sharing! I agree that the pandas -> numpy route is not perfect in terms of performance and memory usage. However, the original C API implementation had a lot of missing features and it was unclear to us how to fill them. As seen in the PR, it's rare that we can add features by removing code. I have some familiarity with the internal of arrow due to a different project, the way that it chunks data with multiple layers of buffer is not easy for a traditional ML library (like XGBoost and most of the scikit-learn implemented algorithms) to handle.

**Comment by trivialfis** [comment]
We have changed the arrow input to use a per-column-based format. However, the performance is still not great due to concatenating the dataframe, which is necessary for gradient boosting. The pandas dataframe is processed faster simply because it uses contiguous memory for each column. I think further optimization is out of scope for XGBoost, but feel free to raise new issues if there are suggestions on how to speedup the process.
